### PR TITLE
[Fix] Resolve multi-dim scalar BufferLoad index in T.tile binary ops (#1679)

### LINKED ISSUE #1679: [Bug] T.tile binary ops: multi-dim BufferLoad S[1,3] silently uses flat index 1 (S[0,1])
## Bug Description

`T.tile.max` (and every other binary op sharing `binary_op`) silently ignores all but the first index when a **multi-dimensional** buffer element (`s_ub[1, 3]`) is passed as the src1 scalar. The op behaves as if the value were `s_ub.flatten()[1]` (i.e. `s_ub[0, 1]`) instead of `s_ub[1, 3]`.

Root cause: `binary_op` in `tilelang/language/ascend_tile.py` forwards only `indices_1[0]` to the codegen (`GetValue(indices[0])`), so any index beyond the first is dropped without an error.

This is a user-facing footgun: `dst[i] = max(src0[i], S[1,3])` silently computes `max(src0[i], S.flatten()[1])`.

## Environment

**Hardware:**
- NPU Device: Ascend 910B3

**Software:**
- OS: Ubuntu (container)
- CANN version: 8.5.2
- torch-npu version: torch_npu (with torch 2.x)
- Python version: 3.11
- TileLang-Ascend version: baseline `bcd9aeb5` + doc/test-only branches (no front-end changes)
- Installation method: build from source

## Code to Reproduce

```python
import torch
import tilelang
import tilelang.language as T

PASS_CONFIGS = {
    tilelang.PassConfigKey.TL_ASCEND_AUTO_CV_COMBINE: True,
    tilelang.PassConfigKey.TL_ASCEND_AUTO_CV_SYNC: True,
    tilelang.PassConfigKey.TL_ASCEND_AUTO_SYNC: True,
    tilelang.PassConfigKey.TL_ASCEND_MEMORY_PLANNING: True,
}


@T.prim_func
def kernel(A: T.Tensor((4, 64), "float16"), S: T.Tensor((2, 8), "float16"), C: T.Tensor((4, 64), "float16")):
    with T.Kernel(1, is_npu=True) as (cid, _):
        a_ub = T.alloc_ub((4, 64), "float16")
        s_ub = T.alloc_ub((2, 8), "float16")
        c_ub = T.alloc_ub((4, 64), "float16")
        T.copy(A, a_ub)
        T.copy(S, s_ub)
        T.tile.max(c_ub, a_ub, s_ub[1, 3])  # intended scalar: S[1,3]
        T.copy(c_ub, C)


if __name__ == "__main__":
    torch.manual_seed(0)
    a = torch.randn(4, 64, dtype=torch.float16)
    s = torch.randn(2, 8, dtype=torch.float16)

    func = tilelang.compile(kernel, out_idx=[-1], pass_configs=PASS_CONFIGS, target="ascendc")
    out = func(a.npu(), s.npu())
    torch.npu.synchronize()

    ref_first_index = torch.max(a, s.view(-1)[1])  # what the code actually used
    ref_true_index = torch.max(a, s[1, 3])         # what was requested
    print("out == max(a, S.flatten()[1]):", torch.allclose(out.cpu(), ref_first_index))
    print("out == max(a, S[1,3]):         ", torch.allclose(out.cpu(), ref_true_index))
```

## Error Message

No error is raised. Both backends silently produce the wrong result:

```
[ascendc] out == max(a, S.flatten()[1]) (first index only): True
[ascendc] out == max(a, S[1,3]) (asked index):               False
[ascendc] S.flatten()[1] = 1.7646, S[1,3] = 0.6372
[pto]     out == max(a, S.flatten()[1]) (first index only): True
[pto]     out == max(a, S[1,3]) (asked index):               False
```

## Expected Behavior

- Either correctly resolve the multi-dimensional element access (`s_ub[1, 3]` -> flat index `1*8+3`), or
- Raise a clear error at trace time stating that only 1D single-index BufferLoad is supported (mirroring the documented constraint for `T.tile.max/min`).

## Actual Behavior

`S[1, 3]` silently uses the flattened index `1` (element `S[0, 1]`); results are wrong without any warning. Same behavior on both `ascendc` and `pto` backends, for every binary op sharing `binary_op` (add/sub/mul/div/max/min/bitwise_* etc.).

## Additional Context

**Programming Mode:**
- [x] Developer mode (automatic sync/buffer management)
- [ ] Expert mode (manual T.Scope/T.barrier_all)
- [ ] Mixed mode

**Backend:**
- [x] Ascend C & PTO backend
- [ ] AscendNPU IR backend

**Pass Configs (if applicable):**
```python
pass_configs = {
    tilelang.PassConfigKey.TL_ASCEND_AUTO_CV_COMBINE: True,
    tilelang.PassConfigKey.TL_ASCEND_AUTO_CV_SYNC: True,
    tilelang.PassConfigKey.TL_ASCEND_AUTO_SYNC: True,
    tilelang.PassConfigKey.TL_ASCEND_MEMORY_PLANNING: True,
}
```

**Operator Type:**
Element-wise vector binary ops (`T.tile.max/min`; shared front-end `binary_op` affects `T.tile.add/sub/mul/div/max/min` and bitwise ops).

**Additional Information:**
- Have you checked the [Programming Guide](../docs/TileLang-Ascend%20Programming%20Guide.md)? Yes
- Have you searched existing issues? Yes — no existing issue covers multi-dim BufferLoad index dropping. #1265 (slice producing BufferLoad) and #1351 (bounded-slice parse error) are related but different.
- Any other context or screenshots? `binary_op` at `tilelang/language/ascend_tile.py` (see `indices_1[0]`).

### PR DESCRIPTION
**Pull request by 2Sheldonlaw**
Fixes #1679

## Root cause

本 issue 实为**两层 bug 叠加**，issue 报告的是第一层；只修第一层结果依然是错的。

### 第一层（issue 报告的）：前端丢弃多维索引

`tilelang/language/ascend_tile.py` 的 `binary_op`（`add/sub/mul/div/max/min/bitwise_*` 共用）
在 `src1` 为 BufferLoad 时只转发 `indices_1[0]`：

```python
return T.call_intrin(..., buffer_1.access_ptr("r"),
                     indices_1[0],   # ← s_ub[1, 3] 只剩 1，"3" 被静默丢弃
                     size_0)
```

两个后端 codegen（AscendC `AddsAndMulsOpCodegen` / PTO `BinaryVecOpsCodegen`）都把该参数直接
用作 `GetValue(index)`，于是 `T.tile.max(c_ub, a_ub, s_ub[1, 3])` 生成
`s_ub.GetValue(1)` → 读到 `S.flatten()[1] = S[0][1]`，静默出错。同一模式还存在于 `compare`
（`compare_scalar`）与 `select` 的 BufferLoad 分支。

### 第二层（验证时发现）：稠密扁平偏移 ≠ UB 物理布局

把偏移改成"正确的"稠密值 `1×8+3 = 11` 后**结果仍错**。上板做索引探针实验（(4,8) fp16 UB
buffer，逐个验证 `GetValue(N)` 实际读到的元素）：

| 请求 | 稠密偏移 | 实际读到 |
|---|---|---|
| `s_ub[0,3]` | 3 | `S[0][3]` ✓ |
| `s_ub[1,3]` | 11 | **pad 区垃圾** ✗ |
| `s_ub[2,3]` | 19 | **`S[1][3]`** ✓ |
| `s_ub[2,0]` | 16 | **`S[1][0]`** ✓ |

即 **S 的第 r 行物理上存放在元素 `16r` 处，而非稠密布局的 `8r`**。原因是系统级的既有契约：
`shared.ub` 窄行 tile 的每一行物理 padding 到 32 字节——

- AscendC：`copy_gm_to_ub` 按 DataCopyPad 逐行 burst，MTE burst 粒度使每个不足 32B 的行各占一个 32B 槽；
- PTO：显式建模为 `TileUbDataND<half, 2, 16, 2, 8>`（物理列 16 = 32B/2B，valid 列 8，pad 区填 `PadValue`）；
- `Flatten2DBuffer` pass 对 `shared.ub` 做 256bit 内维对齐（`kScopeForAlignment`），UB 内存规划器按 padding 后的形状分配（(2,8) fp16 实际分到 64B）。

**唯一没遵守这个契约的就是前端标量偏移的计算**（TIR buffer 的 `offset_of` 是稠密行主序）。
因此 `s_ub[1,3]` 的正确偏移是 `1×16+3 = 19`，实测 `GetValue(19)` 精确返回 `S[1][3]`。

## What this PR does

| 文件 | 改动 |
|---|---|
| `tilelang/language/ascend_tile.py` | 新增 `_scalar_load_offset()`：多维标量 BufferLoad 按 **32B padding 行 stride** 解析为物理元素偏移（`s_ub[1,3]` 于 (2,8) fp16 → 19）；1-D buffer、内维为 1 的 buffer（PTO 以 `[1, M]` DN tile 平铺存储，稠密偏移正确）、非 UB buffer 保持稠密偏移——`scale_ub[h_i]` 等 1-D 符号索引用法 byte 级不变。`binary_op` / `compare` / `select` 三处 `indices_1[0]` 全部替换 |
| `src/target/codegen_ascend_pto.cc` | PTO `CompareScalarCodegen` 补上 buffer+index 形式的标量支持（该形式此前直接 `Unresolved call Op(tir.tvm_access_ptr)` 崩溃，属预先存在的缺陷，修复前后 arg 布局相同）；实现镜像同文件 `BinaryVecOpsCodegen` 的 `GetValue` 提取模式 |
| `testing/python/language/test_tilelang_ascend_language_elementwise.py` | 回归测试 ×2 组：`test_binary_scalar_from_multidim_buffer`（max）/ `test_compare_scalar_from_multidim_buffer`（compare+packed mask），各参数化 fp16/fp32 × ascendc/pto。fp16 走 padding stride 路径、fp32 走稠密路径；用 `arange` 保证 `s[1,3] ≠ s.flatten()[1]`，并显式断言结果不等于"first-index"参考 |

## Test results

Ascend910B3 / CANN 9.0.0 / 本地构建（commit `a61eeee9`，基于 `upstream/ascendc_pto` `0e9a1a5b` rebase，rebase 后重建并复验通过）。

| 验证 | 结果 |
|---|---|
| issue 原始复现脚本（ascendc + pto） | 修复前 `out == max(a, S.flatten()[1]): True`；修复后双后端 `out == max(a, S[1,3]): True`，且 `flatten()[1]: False` |
| 新增回归测试 | 8/8 通过（含修复前后 A/B 语义：`assert not allclose(out, ref_first_index)`） |
| 既有回归：elementwise 全量 | 492 passed / 2 xfailed |
| 既有回归：tail_mask_codegen + tail_block / select + explicit_tmp / vid_reduction（含 `scale_ub[h_i]` 1-D 符号索引路径） | 316 / 76 / 17 全部通过 |
| `ruff check` / `clang-format --dry-run -Werror` | 干净 |

## ⚠️ 遗留的更深层次问题（本 PR 未修，建议单独跟踪）

1. **窄行 2-D UB buffer 的整块 elementwise 算子**：`T.tile.exp` 等以稠密元素数（`prod(shape)`）
   作为操作长度，而物理布局是行 padding 的——第二行起读到 pad 区。实测 `exp` 在 (2,8) fp16
   buffer 上第二行输出负值（exp 不可能为负）。这是同一布局契约的系统性矛盾（本 PR 只修标量
   读取路径），建议另立 issue。
2. **`T.tile.select` 的 BufferLoad 分支**：AscendC `SelectCodegen` 的 `src1_type==0` 分支把
   `args[4]`（access_ptr）同时当作 buffer 指针和 GetValue 下标（下标实际在 `args[5]`），
   编译期即 `TVMError`。已实测修复前后行为相同（都崩），属预先存在的独立 codegen 缺陷；
   本 PR 修了该路径的前端偏移计算，未动其 codegen。
3. **(M,1) 内维为 1 的 UB buffer**：PTO 平铺为 `[1, M]` DN tile（稠密偏移正确），AscendC
   逐行 burst 会按 32B 槽摆放——两后端物理布局可能不一致。本 PR 对内维为 1 的 buffer 保持
   稠密偏移（与现状一致，零行为变化），该潜在分歧留待后续核查。

## Reproduce snippet（issue 原始复现，修复前）

```python
@T.prim_func
def kernel(A: T.Tensor((4, 64), "float16"), S: T.Tensor((2, 8), "float16"), C: T.Tensor((4, 64), "float16")):
    with T.Kernel(1, is_npu=True) as (cid, _):
        a_ub = T.alloc_ub((4, 64), "float16")
        s_ub = T.alloc_ub((2, 8), "float16")
        c_ub = T.alloc_ub((4, 64), "float16")
        T.copy(A, a_ub)
        T.copy(S, s_ub)
        T.tile.max(c_ub, a_ub, s_ub[1, 3])  # intended scalar: S[1,3]
        T.copy(c_ub, C)
```

修复前（两后端一致）：

```text
out == max(a, S.flatten()[1]): True   # 实际用了 1.7646 (S[0][1])
out == max(a, S[1,3]):          False # 用户要的 0.6372 没用上
```

修复后（两后端一致）：

```text
out == max(a, S.flatten()[1]): False
out == max(a, S[1,3]):          True
```

### CODE REVIEW

**Review by ChaoyangJi — APPROVED** [review]
lgtm
